// @account-kit/react/dist/esm/components/auth/card/passkey.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { http, HttpResponse } from "msw";
import { useEffect } from "react";
import { useUiConfig } from "../../../hooks/useUiConfig.js";
import { useAuthContext } from "../context.js";
import { AuthCard } from "./index.jsx";
const PasskeyStory = (props) => {
    const { updateConfig } = useUiConfig(({ updateConfig }) => ({
        updateConfig,
    }));
    let sections = [[{ type: "passkey" }]];
    const ui = {
        theme: "dark",
        primaryColor: {
            light: "red",
            dark: "#9AB7FF",
        },
        borderRadius: "sm",
        illustrationStyle: "outline",
        logoLight: undefined,
        logoDark: undefined,
    };
    useEffect(() => {
        const uiConfig = {
            auth: {
                showEmail: true,
                showExternalWallets: false,
                showPasskey: true,
                addPasskey: true,
                sections,
            },
            ui,
            supportUrl: "#",
        };
        updateConfig(uiConfig);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.authType]);
    const { setAuthStep } = useAuthContext();
    useEffect(() => {
        setAuthStep({ type: "passkey_create" });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    return _jsx(AuthCard, { ...props });
};
const meta = {
    title: "Passkey",
    component: PasskeyStory,
    args: {
        authType: "passkey",
    },
    parameters: {
        msw: {
            handlers: [
                http.post("/api/rpc/signer/v1/lookup", () => {
                    return HttpResponse.json({
                        orgId: "483c1263-a6e3-4db1-a8b0-894e4902e404",
                    });
                }),
            ],
        },
    },
};
export default meta;
export const Default = {};
