// @account-kit/react-native-signer/lib/commonjs/signer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RNAlchemySigner = RNAlchemySigner;
exports.RNAlchemySignerSingleton = void 0;
var _signer = require("@account-kit/signer");
var _zod = require("zod");
var _client = require("./client.js");
const RNAlchemySignerParamsSchema = _zod.z.object({
  client: _zod.z.custom().or(_client.RNSignerClientParamsSchema)
}).extend({
  sessionConfig: _signer.SessionManagerParamsSchema.omit({
    client: true,
    storage: true
  }).optional()
});
class RNAlchemySignerSingleton extends _signer.BaseAlchemySigner {
  constructor(params) {
    if (!!RNAlchemySignerSingleton.instance) {
      return RNAlchemySignerSingleton.instance;
    }
    const {
      sessionConfig,
      ...params_
    } = RNAlchemySignerParamsSchema.parse(params);
    let client;
    if ("connection" in params_.client) {
      client = new _client.RNSignerClient(params_.client);
    } else {
      client = params_.client;
    }
    super({
      client,
      sessionConfig
    });
    this.signerType = "rn-alchemy-signer";
  }
  static getInstance(params) {
    if (!this.instance) {
      this.instance = new RNAlchemySignerSingleton(params);
    }
    return this.instance;
  }
}

/**
 * Factory function to create or retrieve a singleton instance of RNAlchemySigner.
 *
 * @example
 * ```ts twoslash
 * import { RNAlchemySigner } from "@account-kit/react-native-signer";
 *
 * const signer = RNAlchemySigner({
 *  client: {
 *    connection: {
 *      apiKey: "YOUR_API_KEY"
 *    },
 *  },
 *  // optional config to override default session manager configs
 *  sessionConfig: {
 *    expirationTimeMs: 1000 * 60 * 60 // 60 minutes
 *  }
 * });
 * ```
 *
 * @param {RNAlchemySignerParams} params The parameters required to configure the RNAlchemySigner instance.
 * @returns {RNAlchemySignerSingleton} The singleton instance of RNAlchemySigner configured with the provided parameters.
 */
exports.RNAlchemySignerSingleton = RNAlchemySignerSingleton;
function RNAlchemySigner(params) {
  const instance = RNAlchemySignerSingleton.getInstance(params);
  return instance;
}
